model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java

```diff
@@ -0,0 +1,308 @@
+        }
+        
+        return groupModelStream.count();
+    }
+
+    @Override
+    public Long getGroupsCountByNameContaining(RealmModel realm, String search) {
+        return searchForGroupByNameStream(realm, search, null, null).count();
+    }
+
+    @Override
+    public Stream<GroupModel> getGroupsByRoleStream(RealmModel realm, RoleModel role, Integer firstResult, Integer maxResults) {
+        LOG.tracef("getGroupsByRole(%s, %s, %d, %d)%s", realm, role, firstResult, maxResults, getShortStackTrace());
+        Stream<GroupModel> groupModelStream = getGroupsStream(realm).filter(groupModel -> groupModel.hasRole(role));
+
+        if (firstResult != null && firstResult > 0) {
+            groupModelStream = groupModelStream.skip(firstResult);
+        }
+        
+        if (maxResults != null && maxResults > 0) {
+            groupModelStream = groupModelStream.limit(maxResults);
+        }
+
+        return groupModelStream;
+    }
+
+    @Override
+    public Stream<GroupModel> getTopLevelGroupsStream(RealmModel realm) {
+        LOG.tracef("getTopLevelGroupsStream(%s)%s", realm, getShortStackTrace());
+        return getGroupsStream(realm)
+                .filter(groupModel -> Objects.isNull(groupModel.getParentId()));
+    }
+
+    @Override
+    public Stream<GroupModel> getTopLevelGroupsStream(RealmModel realm, Integer firstResult, Integer maxResults) {
+        Stream<GroupModel> groupModelStream = getTopLevelGroupsStream(realm);
+        
+        if (firstResult != null && firstResult > 0) {
+            groupModelStream = groupModelStream.skip(firstResult);
+        }
+        
+        if (maxResults != null && maxResults > 0) {
+            groupModelStream = groupModelStream.limit(maxResults);
+        }
+        
+        return groupModelStream;
+        
+    }
+
+    @Override
+    public Stream<GroupModel> searchForGroupByNameStream(RealmModel realm, String search, Integer firstResult, Integer maxResults) {
+        LOG.tracef("searchForGroupByNameStream(%s, %s, %d, %d)%s", realm, search, firstResult, maxResults, getShortStackTrace());
+        Stream<GroupModel> groupModelStream = getGroupsStream(realm)
+                .filter(groupModel -> groupModel.getName().contains(search));
+
+        if (firstResult != null) {
+            groupModelStream = groupModelStream.skip(firstResult);
+        }
+
+        if (maxResults != null) {
+            groupModelStream = groupModelStream.limit(maxResults);
+        }
+
+        return groupModelStream;
+    }
+
+    @Override
+    public GroupModel createGroup(RealmModel realm, String id, String name, GroupModel toParent) {
+        LOG.tracef("createGroup(%s, %s, %s, %s)%s", realm, id, name, toParent, getShortStackTrace());
+        final UUID entityId = id == null ? UUID.randomUUID() : UUID.fromString(id);
+
+        // Check Db constraint: uniqueConstraints = { @UniqueConstraint(columnNames = {"REALM_ID", "PARENT_GROUP", "NAME"})}
```

Comment:
I don't understand the comment, could you please explain?

Reply:
The thing here is, that in the database we have a constraint that checks there are no two groups with the same name and parent. Here, we need to do it manually.

Reply:
This note has to be present in `GroupProvider.createGroup` JavaDocs